browser/brave_ads/request_ads_enabled_api_browsertest.cc

```diff
@@ -0,0 +1,306 @@
+        base::BindRepeating(check_load_is_rewards_panel, &popup_contents));
+
+    EXPECT_TRUE(content::ExecJs(contents, kBraveRequestAdsEnabled,
+                                content::EXECUTE_SCRIPT_NO_RESOLVE_PROMISES));
+
+    // Wait for the popup to load
+    popup_observer.Wait();
+    rewards_browsertest_util::WaitForElementToAppear(
+        popup_contents, "[data-test-id='rewards-onboarding-main-button']");
+
+    return popup_contents;
+  }
+
+  net::EmbeddedTestServer* https_server() { return https_server_.get(); }
+
+ protected:
+  base::test::ScopedFeatureList feature_list_;
+
+ private:
+  std::unique_ptr<net::EmbeddedTestServer> https_server_;
+  rewards_browsertest::RewardsBrowserTestResponse rewards_response_;
+};
+
+class RequestAdsEnabledApiTestEnabled : public RequestAdsEnabledApiTestBase {
+ public:
+  RequestAdsEnabledApiTestEnabled() {
+    feature_list_.InitAndEnableFeature(
+        brave_ads::features::kRequestAdsEnabledApi);
+  }
+};
+
+IN_PROC_BROWSER_TEST_F(RequestAdsEnabledApiTestEnabled, AdsAlreadyEnabled) {
+  brave_ads::AdsService* ads_service =
+      brave_ads::AdsServiceFactory::GetForProfile(browser()->profile());
+  ads_service->SetEnabled(true);
+
+  GURL url = https_server()->GetURL(kAllowedDomain, "/simple.html");
+
+  ui_test_utils::NavigateToURL(browser(), url);
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+  EXPECT_EQ(url, contents->GetVisibleURL());
+  EXPECT_EQ(true, content::EvalJs(contents, kBraveRequestAdsEnabledExists));
+
+  EXPECT_TRUE(content::ExecJs(contents, kBraveRequestAdsEnabled,
+                              content::EXECUTE_SCRIPT_NO_RESOLVE_PROMISES));
+  EXPECT_EQ(true, content::EvalJs(contents, kResolveRequestAdsEnabledPromise));
+}
+
+IN_PROC_BROWSER_TEST_F(RequestAdsEnabledApiTestEnabled,
+                       PrivateAdsPopupAccepted) {
+  GURL url = https_server()->GetURL(kAllowedDomain, "/simple.html");
+
+  auto* rewards_service = static_cast<brave_rewards::RewardsServiceImpl*>(
+      brave_rewards::RewardsServiceFactory::GetForProfile(
+          browser()->profile()));
+
+  rewards_browsertest_util::StartProcess(rewards_service);
+  rewards_browsertest_util::CreateWallet(rewards_service);
+  rewards_browsertest_util::SetOnboardingBypassed(browser(), true);
+
+  ui_test_utils::NavigateToURL(browser(), url);
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+  EXPECT_EQ(url, contents->GetVisibleURL());
+  EXPECT_EQ(true, content::EvalJs(contents, kBraveRequestAdsEnabledExists));
+
+  content::WebContents* popup_contents = OpenRequestAdsEnabledPopup(contents);
+
+  rewards_browsertest_util::WaitForElementThenClick(
+      popup_contents, "[data-test-id='rewards-onboarding-main-button']");
+
+  EXPECT_EQ(true, content::EvalJs(contents, kResolveRequestAdsEnabledPromise));
+
+  // Wait for ads enabled.
+  WaitForPrefValue(browser()->profile()->GetPrefs(), ads::prefs::kEnabled,
```

Comment:
you shouldn't wait for this, it should already be `true` when `kResolveRequestAdsEnabledPromise` resolves